Coordinated Spawns (Purple Edition)

MrPurple6411 opened the pull request:
### Changes made in this pull request
  - Moved spawn creation to OnBatchFullyLoaded patch
  - Preload unspawned prefabs during initialization postfix to ensure speedy spawning later.

MrPurple6411 committed:
Coordinated Spawns (Purple Edition)

MrPurple6411 committed:
Don't force LWE requirement.

Metious commented on Nautilus/MonoBehaviours/EntitySpawner.cs:
```diff
@@ -35,40 +35,18 @@ IEnumerator SpawnAsync()
-        yield return new WaitUntil(() => lw != null && lw.streamer.globalRoot != null); // need to make sure global root is ready too for global spawns.
-            
-        lw.streamer.cellManager.RegisterEntity(obj);
+        if (lwe != null)
```
Don't make having a `LargeWorldEntity` optional because the game has no idea how to save it and we cant assume that they're transient.

If an object is missing an LWE component, it's most likely that the user forgot to add it, so it's better to let them know about than to have an arbitrary fallback.

Metious reviewed (changes requested):
Testing this pull request, I'm now getting an NRE first that goes like the following:
```
[Error  : Unity Log] NullReferenceException
Stack trace:
CellManager.RegisterCellEntity (LargeWorldEntity ent) (at <9ec6c1b62c324758b9c2d79021505be6>:0)
CellManager.RegisterEntity (LargeWorldEntity lwe) (at <9ec6c1b62c324758b9c2d79021505be6>:0)
Nautilus.MonoBehaviours.EntitySpawner+<>c__DisplayClass2_0.<SpawnAsync>b__0 () (at <c13f3a3dfa5c494f9fd922a160750b0b>:0)
UnityEngine.WaitUntil.get_keepWaiting () (at <bd0d47c27bd84106afaaecc2c74cdb94>:0)
UnityEngine.CustomYieldInstruction.MoveNext () (at <bd0d47c27bd84106afaaecc2c74cdb94>:0)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at <bd0d47c27bd84106afaaecc2c74cdb94>:0)
```
Then I end up with my object infinitely spawning.
![image](https://github.com/SubnauticaModding/Nautilus/assets/71298690/a9ce7c64-2f93-4a5d-9be4-643737f0adee)

This PR is not done yet. I'm converting it to a draft to prevent accidental merges.

MrPurple6411 committed:
Revert "Don't force LWE requirement."

This reverts commit 7cd473a44fd2812f45aaa34f0ef1cb1b960d20eb.

MrPurple6411 committed:
Combined old with new to correctly allow global and batch spawns.

and hopefully fix the NRE that i still cannot replicate.

Metious committed:
Prefab cache no longer adds objects with IDs set

Metious committed:
Fixed infinite spawning for EntitySpawner

The issue was because the spawned objects was registered while being inactive in the scene.

Metious committed:
Coordinated spawns per-batch instead of per-entity

Metious committed:
Make sure it also works when a spawn is registered in-game

Metious reviewed (approved):
Works fine on my tests

MrPurple6411 committed:
Fix 1 fail = all following stopped.

MrPurple6411 committed:
Delayed Spawns don't stop all progress.

MrPurple6411 committed:
Update EntitySpawner.cs

MrPurple6411 committed:
Corrected positioning mistake.

MrPurple6411 commented on Nautilus/MonoBehaviours/EntitySpawner.cs:
```diff
@@ -1,74 +1,106 @@
+                // Cells aren't ready yet. We have to wait until they are.
+                StartCoroutine(WaitForCellLoaded(lws, prefab, spawnInfo, stringToLog));
+                continue;
+            }
```
@Metious can I get you to check this part out because the game itself Ensures a LWE on any registered object and just uses the default cell level which would be near I think

Metious replied:
It always has been the case, but I dislike it because we're assuming cell levels, which are quite important on how the object loads/saves. I think since we always had them error before, it's better to keep it that way so users still get notified for possible mistakes and also backward compatibility.

Also `LargeWorldStreamer.IsRangeActiveAndBuilt` depends on the clipmap being ready, which is extremely slow and kinda broken especially in the void. In my experience it always returned false no matter how much I waited in the spawning zone, but it only returns true when I teleport to the exact spawn position.
I was using that before switching to just checking for the cells being ready; in my tests in SN cells are almost immediately initialized after the batch is ready for compilation, so if BZ doesn't have the method, they probably don't need them anymore due to that, so maybe keep the cell check for SN only?

MrPurple6411 replied:
Before you actually had a zero sized bounds for some reason and were using it based on the batch center.

I converted this to actually use the same code the game itself uses to check the location
By checking a vector3.one×5 at the spawn location instead of the batch Location. 

I will do some testing for void spawns with my way to see.

Metious replied:
I changed it to batch center only after making the entity spawner per-batch. During my tests I was using the spawn info's position and checked it against 0 size, 3 size and 5 size. All of them returned false when they should be spawning. 0 sized bounds is the most precise because it only checks for the exact spawn position's cells and nothing else.

I will test your code again and report back, but the game never waits for the clipmap to be ready to spawn stuff.

Metious replied:
Just tested your code and can confirm that the issue I mentioned exists.

In the void, before my object that is supposed to spawn at `373 -400 1920` was being spawned when I was at `372 -395 1801`, but with your code I had to swim to the exact spawning position.

The same is true for spawns that are in the crater, but you don't have to go inside the spawning position; just a little bit closer to the spawn position but the spawn will still surprise you have to get awfully close.

MrPurple6411 committed:
Spawn dammit.

MrPurple6411 committed:
remove _delayedSpawns

MrPurple6411 committed:
Should Not exist in main branch!
